pkg/flow/api/types.go

```diff
@@ -0,0 +1,98 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package api
+
+import (
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type Flow interface {
+	Filter(f interface{}) Flow
+	Map(f interface{}) Flow
```

Comment:
how about opting to `func (i interface{}) interface{}` instead `interface`?

Reply:
I would explain "any" type here.

There are mainly two reasons,

1. With `any`, we can use two flavors of methods with or without `context.Context`: `func (i interface{}) interface{}` and `func (ctx context.Context, i interface{}) interface{}`
2. With `any`, we can unify the batch query and streaming processing,

```go
func (flow *batchFlow) Filter(f interface{}) api.Flow {
	if unresolvedCriteria, ok := f.([]*modelv1.Criteria); ok {
		flow.conditionOp = &conditionalFilterOperator{
			f:                  flow,
			unresolvedCriteria: unresolvedCriteria,
		}
	} else {
		flow.appendErr(errors.New("Filter can only accept []*modelv1.Criteria"))
	}
	return flow
}
```

The same reason for `Filter`.

Reply:
> With any, we can use two flavors of methods with or without context.Context: func (i interface{}) interface{} and func (ctx context.Context, i interface{}) interface{}

I want to pass the context parameter all the time. If you noticed that, more and more libs and projects are introducing the context as a mandatory argument.

> With any, we can unify the batch query and streaming processing

`Filter`, `Map`, and `Reduce` are streaming standard functions. Their parameter should be an "action". Passing the `criteria` would confuse the API's users. 

You could use the function to pass `modelv1.Criteria`. If you have performance concerns, the compiler is smart, optimizing this function.

Reply:
> > With any, we can use two flavors of methods with or without context.Context: func (i interface{}) interface{} and func (ctx context.Context, i interface{}) interface{}
> 
> I want to pass the context parameter all the time. If you noticed that, more and more libs and projects are introducing the context as a mandatory argument.
> 
> > With any, we can unify the batch query and streaming processing
> 
> `Filter`, `Map`, and `Reduce` are streaming standard functions. Their parameter should be an "action". Passing the `criteria` would confuse the API's users.
> 
> You could use the function to pass `modelv1.Criteria`. If you have performance concerns, the compiler is smart, optimizing this function.

I'm having a try. The first is easy to be implemented. But the problem occurs within the query context.

Currently, `[]*modelv1.Criteria` is used to build the `tsdb.ItemFilter`. With the parameter type `func(context.Context, any) bool`, it is impossible to return `[]*modelv1.Criteria`

Reply:
Found an ugly way out. Anyway, we may improve in the future.

Reply:
If that, a new method is better here. Why could you insist reuse `Filter`?

Reply:
Batch API has been removed